File: customer-service/src/main/java/com/wks/servicemarketplace/customerservice/config/AuthenticationFilter.java
package com.wks.servicemarketplace.customerservice.config;

import com.wks.servicemarketplace.common.auth.Authentication;
import com.wks.servicemarketplace.common.auth.TokenValidator;
import com.wks.servicemarketplace.common.errors.CoreException;
import com.wks.servicemarketplace.customerservice.adapters.auth.DefaultSecurityContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import javax.ws.rs.container.ContainerRequestContext;
import javax.ws.rs.container.ContainerRequestFilter;
import javax.ws.rs.container.PreMatching;
import javax.ws.rs.core.Response;

@PreMatching
public class AuthenticationFilter implements ContainerRequestFilter {

    private static final Logger LOGGER = LoggerFactory.getLogger(AuthenticationFilter.class);
    @Inject
    private TokenValidator tokenValidator;

    @Override
    public void filter(ContainerRequestContext requestContext) {
        final String authorization = requestContext.getHeaderString("Authorization");

        if (authorization == null || authorization.isEmpty()) {
            LOGGER.info("authorization token not found");
            return;
        }

        final String token = authorization.substring("Bearer".length()).trim();

        Authentication authentication = null;
        try {
            authentication = tokenValidator.authenticate(token);
        } catch (CoreException e) {
            requestContext.abortWith(Response.status(Response.Status.UNAUTHORIZED).entity(e.getMessage()).build());
            return;
        }

        requestContext.setSecurityContext(new DefaultSecurityContext(
                authentication,
                false,
                "Bearer"
        ));
        LOGGER.info("Security Context set for user: '{}'", authentication.getName());
    }
}

File: service-provider-service/src/main/kotlin/com/wks/servicemarketplace/serviceproviderservice/config/AuthenticationFilter.kt
package com.wks.servicemarketplace.serviceproviderservice.config

import com.wks.servicemarketplace.common.auth.TokenValidator
import com.wks.servicemarketplace.common.errors.CoreException
import com.wks.servicemarketplace.common.http.httpStatusCode
import com.wks.servicemarketplace.serviceproviderservice.adapters.auth.DefaultSecurityContext
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import javax.inject.Inject
import javax.ws.rs.container.ContainerRequestContext
import javax.ws.rs.container.ContainerRequestFilter
import javax.ws.rs.core.Response

class AuthenticationFilter @Inject constructor(private val tokenValidator: TokenValidator) : ContainerRequestFilter {

    private val LOGGER: Logger = LoggerFactory.getLogger(AuthenticationFilter::class.java)

    override fun filter(requestContext: ContainerRequestContext) {
        try {
            requestContext.getHeaderString("Authorization")
                    ?.takeIf { it.isNotBlank() }
                    ?.substring("Bearer ".length)
                    ?.let {
                        tokenValidator.authenticate(it)
                    }?.also {
                        requestContext.securityContext = DefaultSecurityContext(it, false, "Bearer")
                        LOGGER.info("Security Context set for user: '{}'", it.name)
                    } ?: LOGGER.info("authorization token not found")
        } catch (e: CoreException) {
            requestContext.abortWith(Response.status(e.errorType.httpStatusCode()).entity(e.message).build())
        }
    }
}
